dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,278 @@
+            UpsertOperationName,
+            () => index.Upsert([vector], indexNamespace, cancellationToken)).ConfigureAwait(false);
+
+        return vector.Id;
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> UpsertBatchAsync(
+        IEnumerable<TRecord> records,
+        UpsertRecordOptions? options = default,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(records);
+
+        var indexNamespace = this.GetIndexNamespace();
+
+        var index = await this.GetIndexAsync(this.CollectionName, cancellationToken).ConfigureAwait(false);
+
+        var vectors = VectorStoreErrorHandler.RunModelConversion(
+            DatabaseName,
+            this.CollectionName,
+            UpsertOperationName,
+            () => records.Select(this._mapper.MapFromDataToStorageModel).ToList());
+
+        await this.RunOperationAsync(
+            UpsertOperationName,
+            () => index.Upsert(vectors, indexNamespace, cancellationToken)).ConfigureAwait(false);
+
+        foreach (var vector in vectors)
+        {
+            yield return vector.Id;
+        }
+    }
+
+    private async Task<T> RunOperationAsync<T>(string operationName, Func<Task<T>> operation)
+    {
+        try
+        {
+            return await operation.Invoke().ConfigureAwait(false);
+        }
+        catch (RpcException ex)
+        {
+            throw new VectorStoreOperationException("Call to vector store failed.", ex)
+            {
+                VectorStoreType = DatabaseName,
+                CollectionName = this.CollectionName,
+                OperationName = operationName
+            };
+        }
+    }
+
+    private async Task RunOperationAsync(string operationName, Func<Task> operation)
+    {
+        try
+        {
+            await operation.Invoke().ConfigureAwait(false);
+        }
+        catch (RpcException ex)
+        {
+            throw new VectorStoreOperationException("Call to vector store failed.", ex)
+            {
+                VectorStoreType = DatabaseName,
+                CollectionName = this.CollectionName,
+                OperationName = operationName
+            };
+        }
+    }
+
+    private async Task<Sdk.Index<GrpcTransport>> GetIndexAsync(string indexName, CancellationToken cancellationToken)
+    {
+        // Check the local cache first, if not found create a new one.
+        if (!this._indexMap.TryGetValue(indexName, out Sdk.Index<GrpcTransport>? client))
+        {
+            client = await this._pineconeClient.GetIndex(indexName, cancellationToken).ConfigureAwait(false);
+            this._indexMap[indexName] = client;
+        }
+
+        return client;
+    }
+
+    // TODO: Should namespace information be stored in PineconeVectorRecordStoreOptions, as part of collection name (e.g.: myindex.mynamespace), or not supported at all
+    private string? GetIndexNamespace()
+        => null;
```

Comment:
If this is useful configuration on Pinecone side, I think we should support it. We can put `IndexNamespace` property in `PineconeVectorRecordStoreOptions` as a default one, and we can put it in each Options class for specific operation, like `UpsertRecordOptions`, `DeleteRecordOptions` etc.

Reply:
`GetRecordOptions`, `UpsertRecordOptions`, `DeleteRecordOptions` are not extensible, and my understanding is that we want to keep it that way. So the only realistic knob we have is in `PineconeVectorStoreRecordCollectionOptions` that we pass in the ctor to `PineconeVectorStoreRecordCollection`, where we can pass a default namespace, used for every operation. In the SDK (and in the Pinecone APIs), index is a per-operation setting, so it's a bit awkward for it to be fixed thing in SK. 
However, it is better than before, where VectorRecordStore could actually specify index names per operation, but the namespace would have been fixed across the board. Now I guess users who want to work within several namespaces will create one PineconeVectorStoreRecordCollection per different namespace and use them as needed. Not the cleanest approach, but probably an ok workaround.

Reply:
> `GetRecordOptions`, `UpsertRecordOptions`, `DeleteRecordOptions` are not extensible at a connector level, and my understanding is that we want to keep it that way.

My fault, I forgot that these Options are used for all connectors. We definitely don't want to put Pinecone-related settings into these classes. In this situation, `PineconeVectorStoreRecordCollection` per different namespace looks reasonable to me. We can also add support for namespace later, if it will be needed.

Reply:
ended up adding support for namespaces in this pr